arroyo-state/src/tables.rs

```diff
@@ -316,16 +325,23 @@ impl<K: Key, V: Data> KeyTimeMultiMapCache<K, V> {
                 from_micros(min_watermark - table_descriptor.retention_micros)
             });
 
-        for (timestamp, key, value) in backing_store.get_data_triples(table).await {
+        for (timestamp, key, value, operation) in backing_store.get_data_triples(table).await {
             if timestamp < min_valid_time {
                 continue;
             }
-            values
-                .entry(key)
-                .or_default()
-                .entry(timestamp)
-                .or_default()
-                .push(value);
+            match operation {
+                DataOperation::Insert => {
+                    values
+                        .entry(key)
+                        .or_default()
+                        .entry(timestamp)
+                        .or_default()
+                        .push(value);
+                }
+                DataOperation::Delete => {
+                    values.entry(key).or_default().remove(&timestamp); // TODO: delete just the value?
```

Comment:
For option 4 would I think there are two ways to implement it:
a. extend the `DataOperation::Delete` enum variant to include a timestamp, store that in parquet, and expand that into the individual deletes at checkpoint restoration time.
b. expand the operation into individual deletes at cache update time and write the individual deletes to parquet.

I think option a is more efficient because we only need to do a single linear search of the vector, right?

Reply:
The enum could look something like this:
```
pub enum DataOperation {
    Insert,
    DeleteKey, // delete single key/value if Global/TimeKeyMap, delete all values for key if KeyTimeMultiMap 
    DeleteBefore, // delete all values for key before timestamp (only for KeyTimeMultiMap)
}
```

Reply:
I've described the operations in the [doc](https://docs.google.com/document/d/1430Raxk4jmJSZOhEjNbUGdBex9hCC4UZhyshlPaTaa8/edit), but the initial implementation let's just do `DeleteKey`.